feat: release transform filter (#2651)
kyma-project/telemetry-manager: docs/contributor/otel-bump-guide.md

# OpenTelemetry Dependency Bump Guide

Guide for maintainers and contributors when bumping `opentelemetry-collector` and `opentelemetry-collector-contrib` dependencies.

---

## Pre-bump Checklist

### Review Changelogs

Review the changelogs for both repositories to identify relevant changes:

- [OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-collector/releases)
- [OpenTelemetry Collector Contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib/releases)

Focus on these areas:

> - Breaking changes, bug fixes, and enhancements for:

Reviewer: ```suggestion
Identify breaking changes, bug fixes, and enhancements for the following components:
```

  - Receivers
    - `filelogreceiver` (contrib)
    - `kubeletstatsreceiver` (contrib)
    - `k8sclusterreceiver` (contrib)
    - `otlpreceiver`
    - `prometheusreceiver` (contrib)
  - Processors
    - `batchprocessor`
    - `filterprocessor` (contrib)
    - `k8sattributesprocessor` (contrib)
    - `memorylimiterprocessor`
    - `resourceprocessor` (contrib)
    - `transformprocessor` (contrib)
  - Exporters
    - `otlpexporter`
    - `otlphttpexporter`
  - Extensions
    - `filestorage` (contrib)
    - `healthcheckextension` (contrib)
    - `k8sleaderelector` (contrib)
    - `pprofextension` (contrib)
  - Connectors
    - `forwardconnector`
    - `routingconnector` (contrib)
- **OTTL** (OpenTelemetry Transformation Language) updates
- **Internal metrics** modifications
- **Deprecation notices**

### OTTL Changes

> Check for:

Reviewer: ```suggestion
1. Check whether any functions changed.
```

> #### Function Modifications

Reviewer: ```suggestion
```

- New Functions
- Name changes
- Signature changes
- Function deprecations

> [!IMPORTANT]

> > Processors may define additional OTTL functions which are restricted to specific contexts. The `filterprocessor` introduces [metrics only functions](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/filterprocessor#ottl-functions). If new metric context-specific functions exist, add these to the documentation of functions that can't be used, since we pin context to `datapoint` in MetricPipeline, metrics only functions will not be available for users.

Reviewer: ```suggestion
   The `filterprocessor` may introduce functions that operate on entire metrics (using the `metric` context. However, our MetricPipeline operates on individual data points (a `datapoint` context) and cannot use such [metrics only functions](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/filterprocessor#ottl-functions).
   If you find new functions using the `metric` context, add them to the user documentation of functions that aren't supported by the filterprocessor.
```

### Processor Updates

#### Filter Processor
- Monitor the availability of context inference in `filterprocessor` in this [issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/37904)

#### Transform Processor
- Verify any changes related to context inference

### Internal Metrics

Metrics can often change without notice. See examples below:

Common issues:
- [Prometheus metric name changes](https://github.com/open-telemetry/opentelemetry-collector/issues/13544)
- [Attribute additions/removals](https://github.com/open-telemetry/opentelemetry-collector/issues/9943)

> ### Breaking Changes and Feature Gates

Reviewer: ```suggestion
### Identify and Plan for Breaking Changes
```

- [OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-collector/releases)
- [OpenTelemetry Collector Contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib/releases)

 Focus on the following areas:

### 1. Review Changed Components

Identify breaking changes, bug fixes, and enhancements for the following components:
  - Receivers
    - `filelogreceiver` (contrib)
    - `kubeletstatsreceiver` (contrib)
    - `k8sclusterreceiver` (contrib)
    - `otlpreceiver`
    - `prometheusreceiver` (contrib)
  - Processors
    - `batchprocessor`
    - `filterprocessor` (contrib)
    - `k8sattributesprocessor` (contrib)
    - `memorylimiterprocessor`
    - `resourceprocessor` (contrib)
    - `transformprocessor` (contrib)
  - Exporters
    - `otlpexporter`
    - `otlphttpexporter`
  - Extensions
    - `filestorage` (contrib)
    - `healthcheckextension` (contrib)
    - `k8sleaderelector` (contrib)
    - `pprofextension` (contrib)
  - Connectors
    - `forwardconnector`
    - `routingconnector` (contrib)
- **Deprecation notices**

### 2. Detect OTTL Changes

1. Check whether any functions changed.

- New Functions

> - Name changes

Reviewer: ```suggestion
   - Name changes
```

- Signature changes
- Function deprecations

2. Check for incompatible OTTL function contexts.
   The `filterprocessor` may introduce functions that operate on entire metrics (using the `metric` context. However, our MetricPipeline operates on individual data points (a `datapoint` context) and cannot use such [metrics only functions](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/filterprocessor#ottl-functions).
   If you find new functions using the `metric` context, add them to the user documentation of functions that aren't supported by the filterprocessor.

> ### Review Processor Updates

Reviewer: ```suggestion
### 3. Review Processor Updates
```

#### Filter Processor
- Monitor the availability of context inference in `filterprocessor` in this [issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/37904)

#### Transform Processor
- Verify any changes related to context inference

> ### Check Internal Metrics

Reviewer: ```suggestion
### 4. Check Internal Metrics
```

Metrics can often change without notice. See the following examples:

- [Prometheus metric name changes](https://github.com/open-telemetry/opentelemetry-collector/issues/13544)
- [Attribute additions/removals](https://github.com/open-telemetry/opentelemetry-collector/issues/9943)

> ### Identify and Plan for Breaking Changes

Reviewer: ```suggestion
### 5. Identify and Plan for Breaking Changes
```
